lib/iris/fileformats/netcdf.py

```diff
@@ -1593,57 +1725,158 @@ def _get_dim_names(self, cube):
+            for coord in matching_coords:
+                if coord not in self._dim_names_and_coords.coords:
+                    self._dim_names_and_coords.append(dim_name, coord)
+
+            # Add the latest name to the returned list
+            dimension_names.append(dim_name)
+
+        # Get info on mesh, first.
+        dimension_names.clear()
+        mesh = cube.mesh
+        if mesh is None:
+            cube_mesh_dim = None
+        else:
+            # Identify all the mesh dimensions.
+            # NOTE: one of these will be a cube dimension, but that one does not
+            # get any special handling.  We *do* want to list/create them in a
+            # definite order (face,edge,node), and before non-mesh dimensions.
+            mesh_location_dimnames = {}
+            for location in MESH_LOCATIONS:
+                # Find if this location exists in the mesh, and a characteristic
+                # coordinate to identify it with.
+                # To use only _required_ UGRID components, we use a location
+                # coord for nodes, but a connectivity for faces/edges
+                if location == "node":
+                    # For nodes, identify the dim with a coordinate variable.
+                    # Selecting the X-axis one for definiteness.
+                    dim_coords = mesh.coords(include_nodes=True, axis="x")
                 else:
-                    # Return the dim_name associated with the existing
-                    # coordinate.
-                    dim_name = self._name_coord_map.name(coord)
-                    dimension_names.append(dim_name)
+                    # For face/edge, use the non-optional connectivity variable.
+                    cf_role = f"{location}_node_connectivity"
+                    dim_coords = mesh.connectivities(cf_role=cf_role)
+                if len(dim_coords) > 0:
+                    # As the mesh contains this location, we want to include this
+                    # dim in our returned mesh dims.
+                    # We should have 1 identifying variable (of either type).
+                    assert len(dim_coords) == 1
+                    mesh_coord = dim_coords[0]
+                    match = mesh_coord in self._dim_names_and_coords.coords
+                    if match:
+                        # For mesh-identifying coords, we require the *same*
+                        # coord, not an identical one (i.e. "is" not "==")
+                        name = self._dim_names_and_coords.name(mesh_coord)
+                        stored_coord = self._dim_names_and_coords.coord(name)
+                        match = mesh_coord is stored_coord
+                    if match:
+                        # Use the previous name for this dim of this mesh, from
+                        # a previously saved cube with the same mesh.
+                        dim_name = self._dim_names_and_coords.name(mesh_coord)
```

Comment:
@pp-mo Surely this is simply `dim_name = name`, right?

Reply:
Fixed in the "new world", where we always fetch a match + then see if it is None.